src/cycle_basis.c

```diff
@@ -0,0 +1,660 @@
+ * \return Error code.
+ * 
+ * Time complexity: O(|V| + |E|) (spanning tree) + ?? (backtracking).
+ * 
+ */
+int igraph_fundamental_cycle_basis(const igraph_t *graph,
+    igraph_vector_ptr_t *basis) {
+
+    long int no_of_nodes=igraph_vcount(graph);
+    long int no_of_edges=igraph_ecount(graph);
+    long int no_of_cycles=no_of_edges - (no_of_nodes - 1);
+    igraph_vector_char_t added_nodes, added_edges;
+    igraph_vector_t parent_nodes, parent_edges, dist, inc_edges;
+    igraph_dqueue_t q=IGRAPH_DQUEUE_NULL;
+
+    long int i, j, k;
+
+    // Given a spanning tree (forest), fundamental cycles are given by
+    // unions of:
+    // - a path within the tree
+    // - a single edge connecting end to start nodes outside the tree
+    // The number of fundamental cycles is #edges - (#vertices -1),
+    // because the quantity in parenthesis is the number of edges in a
+    // tree (no cycles).
+    
+    // The key is that every non-tree edge generates a fundamental cycle
+    // so we run a BFS and whenever we encounter a vertex which is not
+    // new we are:
+    // - either trying to walk back the last edge we just walked
+    // - or using a non-tree edge (since both vertices are known already)
+    // Note: every edge must be walked once, so even after the tree
+    // is complete we still have to finish walking the remaining edges.
+    // Note: we cannot use the igraph_bfs function because we have to
+    // inspect outgoing edges that point to vertices that have already
+    // been added
+
+    IGRAPH_VECTOR_CHAR_INIT_FINALLY(&added_nodes, no_of_nodes);
+    IGRAPH_VECTOR_CHAR_INIT_FINALLY(&added_edges, no_of_edges);
+    IGRAPH_VECTOR_INIT_FINALLY(&parent_nodes, no_of_nodes);
+    IGRAPH_VECTOR_INIT_FINALLY(&parent_edges, no_of_edges);
+    IGRAPH_VECTOR_INIT_FINALLY(&dist, no_of_nodes);
+    IGRAPH_DQUEUE_INIT_FINALLY(&q, 100);
+    IGRAPH_VECTOR_INIT_FINALLY(&inc_edges, 0);
+
+    for (i = 0; i < no_of_nodes; i++) {
+        /* Start of a new connected component including the first one */
+        if (added_nodes[i] > 0) { continue; }
+
+        IGRAPH_ALLOW_INTERRUPTION();
+
+	/* This is the root, mark it visited */
+        added_nodes[i] = 1;
+
+	/* The queue stores the next vertex, the edge to it, and its distance to root */
+        IGRAPH_CHECK(igraph_dqueue_push(&q, i));
+        IGRAPH_CHECK(igraph_dqueue_push(&q, -1));
+        IGRAPH_CHECK(igraph_dqueue_push(&q, 0));
+        while (! igraph_dqueue_empty(&q)) {
+          /* current node, incoming edge, and distance from tree root */
+          long int actnode=(long int) igraph_dqueue_pop(&q);
+	  long int actedge=(long int) igraph_dqueue_pop(&q);
+          long int actdist=(long int) igraph_dqueue_pop(&q);
+
+	  /* check all edges connected to this vertex */
+          IGRAPH_CHECK(igraph_incident(graph, &inc_edges, (igraph_integer_t) actnode,
+				       IGRAPH_ALL));
+
+          for (j=0; j<igraph_vector_size(&inc_edges); j++) {
```

Comment:
Please avoid using `vector_size` in `for` loops like this. Instead, get the size first, store it in a variable, then use that variable instead.

Here it would not matter that much because there are already several other function calls in the loop, but in general it's good to get into the habit of storing the size in a variable. In a tight loop, that function call would bring a noticeable performance hit.

Reply:
This is copycat of the exact same pattern in `spanning_trees.c`, e.g. line 321. I think the compiler will put the `igraph_vector_size(&inc_edges)` into a variable anyway, so I'm not sure a human should deal with these kinds of optimizations.

So if this is a performance problem because the compiler does not do its job, I can open a separate PR to fix it everywhere else too. @ntamas @vtraag ?

Reply:
I have been surprised several times in the past by the complexity of optimizations compilers can do automatically, but I'm still unsure :)

On one hand, yes, it should be the compiler's job to optimize this and leave it to us mere mortals to deal with the more interesting parts of programming. On the other hand, in order to do the optimization here in this particular case, the compiler (let it be `gcc`, `clang`, MSVC or something else) must be able to infer that 1) `igraph_vector_size()` is side-effect-free (okay, that's actually quite easy), and that 2) the thing that `igraph_vector_size()` calculates is an invariant of the loop so it can be lifted to an external variable (which might even end up being in a register only). If the compiler can infer this, that's great.

As for me, I routinely lift the size of a vector into a variable on its own and then use that variable in the loop -- compilers were not that good in the past and old habits die hard. Also, it gives me peace of mind to do so because then I can say that it probably works even on a crappy compiler. However, if someone leaves the function call in the loop, that's also okay if I know it's a simple function call that does not do anything complex.

Reply:
The compiler cannot infer that `igraph_vector_size` is side-effect free because it does not have the definition of that function. It is in a different translation unit.

Link-time optimization might be able to mitigate this to some extent, but it's not guaranteed, and LTO does not even work on all platforms. We should not rely on it.

Reply:
@iosonofabio In C++ one can just do `for (int i=0; i < vec.size(); ++i)` because `vec.size()` will be an inline function that lives in the header. The compiler can analyse it and figure out that it does not need to call it for every loop iteration. In C, this is not possible because there is no general support for inline functions and the definition will simply not be visible to the compiler.

It is good practice to make a habit of avoiding this pattern. It's very easy to do so and then you don't have to keep thinking about how big the performance impact is anyway.